### ISSUE TITLE
Type.IsSerializable and FieldInfo.IsNotSerialized are marked with Obsolete in .NET 8

### ISSUE THREAD

**Issue by gerhard17** [dotnet/runtime#102809]
### Description

While SerializableAttribute and NonSerializedAttribute can be applied to types / fields, the corresponding relection APIs are marked as obsolete. 

I understand that formatter-based serialization is obsolete, but this attributes are currently used by a custom implemented serialization (IPC) mechanism. Porting a part of our enterprise infrastructure from .NET Framework to .NET 8 this leads to various obsolete warnings and the probable danger of this APIs will be removed in a future version of .NET.

### Reproduction Steps

Using Type.IsSerializable and FieldInfo.IsNotSerialized in code leads to an obsolete warning: 

Warning	SYSLIB0050	'Type.IsSerializable' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050)

Warning	SYSLIB0050	'FieldInfo.IsNotSerialized' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050)

### Expected behavior

Using Type.IsSerializable and FieldInfo.IsNotSerialized are functional and lead to no obsolete warnings.

### Actual behavior

Using Type.IsSerializable and FieldInfo.IsNotSerialized leads to obsolete warnings.

### Regression?

As far I read the spec the obsolete marking was introduced from .NET 7 to .NET 8

In my case I'm porting from .NET 4.6.2 to .NET 8

### Known Workarounds

Using an own custom attribute set with proper custom reflection support.

But to change all components using SerializableAttribute and NonSerializedAttribute would take a major effort and will cause hard to track down errors during migration phase. Also the problem of porting legacy (bad maintained) components would arise.

### Configuration

.NET4.6.2 / .NET 8, Windows x64 (but not specific to that)

### Other information

Please remove the ObsolteAttribute from Type.IsSerializable and FieldInfo.IsNotSerialized and keep their functionality unchanged.
Also do NOT consider removing SerializeableAttribute and NonSerializedAttribute.

**Comment by jkotas** [comment]
You can disable the warning with the specific warning code if you are not able to fix it right way.

**Comment by gerhard17** [comment]
Yes I know I can remove the warning. But obsolete means the functionality maybe removed in future. 

My suggestion is to keep the functionality.

And when the functionality will remain, then obsolete should be removed from these two methods.

**Comment by osexpert** [comment]
IMO it was a mistake to obsolete ISerializable.GetObjectData(SerializationInfo, StreamingContext) as well.
It is not only used by "Formatter-based serialization" but also eg. [vs-streamjsonrpc](https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/connecting.md) and a few other rpc frameworks.

**Comment by osexpert** [comment]
> > IMO it was a mistake to obsolete ISerializable.GetObjectData(SerializationInfo, StreamingContext) as well.
> 
> The `ISerializable` model is quite poorly designed. The usage of it is really not encouraged.
> 
The obsolete text says its obsolete because its "formatter-based" thou, nothing about poorly designed etc. Imo this obsolete is sailing under false flag.

**Comment by jkotas** [comment]
The link attached to the obsolete message points to document with more details that points to documents with even more details.

Any serializer that is based on the serialized model provided by these APIs is exposed to [the same security risks](https://learn.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide) as the binary formatter. That's why these APIs are marked obsolete.
